macros/parserPopUp.pl

```diff
@@ -90,51 +121,73 @@ sub new {
   my $self = shift; my $class = ref($self) || $self;
   shift if Value::isContext($_[0]); # remove context, if given (it is not used)
   my $choices = shift; my $value = shift;
-  Value::Error("A PopUp's first argument should be a list of menu items")
+  Value->Error("A PopUp's first argument should be a list of menu items")
     unless ref($choices) eq 'ARRAY';
-  Value::Error("A PopUp's second argument should be the correct menu choice")
+  Value->Error("A PopUp's second argument should be the correct menu choice")
     unless defined($value) && $value ne "";
-  my %choice; map {$choice{$_} = 1} @$choices;
-  Value::Error("The correct choice must be one of the PopUp menu items")
-    unless $choice{$value};
   $self = bless {data => [$value], context => $context, choices => $choices}, $class;
+  $self->getChoiceOrder;
+  my %choice; map {$choice{$_} = 1} @{$self->{choices}};
+  if (!$choice{$value}) {
+    my @order = map {ref($_) eq "ARRAY" ? @$_ : $_} @$choices;
+    if ($value =~ m/^\d+$/ && $order[$value]) {$self->{data}[0] = $order[$value]}
+      else {Value->Error("The correct choice must be one of the PopUp menu items")}
+  }
   return $self;
 }
 
 #
+#  Get the choices into the correct order (randomizing where requested)
+#
+sub getChoiceOrder {
+  my $self = shift;
+  my @choices = ();
+  foreach my $choice (@{$self->{choices}}) {
+    if (ref($choice) eq "ARRAY") {push(@choices,$self->randomOrder($choice))}
+      else {push(@choices,$choice)}
+  }
+  $self->{choices} = \@choices;
+}
+sub randomOrder {
+  my $self = shift; my $choices = shift;
+  my %index = (map {$main::PG_random_generator->rand => $_} (0..scalar(@$choices)-1));
+  return (map {$choices->[$index{$_}]} main::PGsort(sub {$_[0] lt $_[1]},keys %index));
+}
+
+#
 #  Create the menu list
 #
 sub menu {shift->MENU(0,@_)}
 sub MENU {
   my $self = shift; my $extend = shift; my $name = shift;
-  my $list = $self->{choices}; my $menu = "";
+  my @list = @{$self->{choices}}; my $menu = "";
   $name = main::NEW_ANS_NAME() unless $name;
   my $answer_value = (defined($main::inputs_ref->{$name}) ? $main::inputs_ref->{$name} : '');
   my $label = main::generate_aria_label($name);
   if ($main::displayMode =~ m/^HTML/) {
-    $menu = qq!<SELECT class="pg-select" NAME="$name" id="$name" aria-label="$label" SIZE=1>\n!;
-    foreach my $option (@$list) {
-      my $selected = ($option eq $answer_value) ? " SELECTED" : "";
-      $menu .= qq!<OPTION$selected VALUE="$option">$option</OPTION>\n!;
+    $menu = qq!<select class="pg-select" name="$name" id="$name" aria-label="$label" size="1" style="max-width:100%">\n!;
```

Comment:
OK, the `max-width` of 95% was causing the longest menu item to be truncated when it was selected.  That was because there was nothing else in the test problem except the menu, and so the container for the menu was just big enough for the menu, and 95% of that was smaller than the menu.  In a real problem, there is likely to be text that makes the problem wider, so I guess it is not a problem, though I still think it is a work-around that can have unexpected consequences.  For example, if you use popups inside a table (popups in their own cells), then they will by 95% of their usual size even on large screens.  I'm wondering if a media-specific CSS setting might be a better solution for small screen issues than making global settings like this.

I will remove the commit that made this change.  I suspect that there will be people who make wide menus by putting in long choice texts, and reducing the size on small screens may be important.